parachain/pallets/ethereum-beacon-light-client/src/lib.rs

```diff
@@ -0,0 +1,562 @@
+					let mut data = [0u8; 64];
+					data[0..32].copy_from_slice(&(value.0));
+					data[32..64].copy_from_slice(&(branch[i as usize].0));
+					value = sha2_256(&data).into();
+				}
+				else {
+					let mut data = [0u8; 64]; // right node
+					data[0..32].copy_from_slice(&(branch[i as usize].0));
+					data[32..64].copy_from_slice(&(value.0));
+					value = sha2_256(&data).into();
+				}	
+			}
+			return value == root
+		}
+
+		//** Helper functions **//
+		// https://github.com/ethereum/consensus-specs/blob/dev/specs/altair/sync-protocol.md#helper-functions
+		fn get_subtree_index(generalized_index: u64) -> u64 {
+			let base: u32 = 2; 
+
+			let floorlog2_finalized_root_index = Self::floorlog2(FINALIZED_ROOT_INDEX);
+
+			generalized_index % base.pow(floorlog2_finalized_root_index as u32) as u64
+		}
+
+		fn get_active_header(update: LightClientUpdate) -> BeaconBlockHeader {
+			// The "active header" is the header that the update is trying to convince us
+			// to accept. If a finalized header is present, it's the finalized header,
+			// otherwise it's the attested header
+			match update.finalized_header {
+				Some(finalized_header) => finalized_header,
+				None => update.attested_header,
+			}
+		}
+
+		fn get_safety_threshold(
+			prev_max_active_participants: u64,
+			curr_max_active_participants: u64,
+		) -> u64 {
+			cmp::max(prev_max_active_participants, curr_max_active_participants)
+		}
+
+		fn get_sync_committee_sum(sync_committee_bits: Vec<u8>) -> u64 {
+			sync_committee_bits.iter().sum::<u8>() as u64
+		}
+
+		/// Return the domain for the domain_type and fork_version.
+		fn compute_domain(domain_type: Vec<u8>, fork_version: Option<Version>, genesis_validators_root: Root) -> Domain {
+			let unwrapped_fork_version: Version;
+			if fork_version.is_none() {
+				unwrapped_fork_version = Self::get_genesis_fork_version();
+			} else {
+				unwrapped_fork_version = fork_version.unwrap();
+			}
+			// TODO this may not be needed because we pass genesis_validators_root from relayer.
+			//if genesis_validators_root is None:
+			//	genesis_validators_root = Root()  # all bytes zero by default
+			
+			let fork_data_root = Self::compute_fork_data_root(unwrapped_fork_version, genesis_validators_root);
+
+			let mut domain = [0u8; 32];
+
+			domain[0..4].copy_from_slice(&(domain_type));
+			domain[4..32].copy_from_slice(&(genesis_validators_root.0));
+
+			domain.into()
+		}
+
+		fn compute_fork_data_root(current_version: Version, genesis_validators_root: Root) -> Root {
+			Self::hash_tree_root(ForkData{
+				current_version: current_version,
+				genesis_validators_root: genesis_validators_root.as_bytes().to_vec(),
+			})
+		}
+
+		fn compute_signing_root(attested_header: BeaconBlockHeader, domain: H256) -> Vec<u8> {
+			todo!()
+		}
+
+		fn bls_fast_aggregate_verify(participant_pubkeys: Vec<Vec<u8>>, signing_root: Vec<u8>, sync_committee_signature: Vec<u8>) -> bool {
+			todo!()
```

Comment:
For BLS signature verification, the reference eth2 implementation (also rust) suggests that the [milagro](https://github.com/apache/incubator-milagro-crypto-rust
) crypto library could be a good fit: https://github.com/sigp/lighthouse/blob/90d5ab15660e62e16390731d92c74b640a70bed9/crypto/bls/src/lib.rs#L12
There may be other good candidates we can work with. However we have a strict constraints to ensure that the code can be compiled down to WASM:
1. Must be written in pure Rust (ie no bindings to C/C++ libraries, inline assembly, etc).
2. Must compile without the standard library available: https://docs.rust-embedded.org/book/intro/no-std.html

If we go ahead with Milagro, we know it supports (1), so we will still need to fork it to support (2).

Reply:
Ah we may be in luck, in this variant/subset of the milagro library supports `no_std` out of the box: https://github.com/sigp/milagro_bls

Big warnings about the lack of security audits though. But I think we can address that later on.